frontend-wasm/src/code_translator/mod.rs

```diff
@@ -0,0 +1,1035 @@
+            hash_map::Entry::Occupied(entry) => *entry.get(),
+            hash_map::Entry::Vacant(entry) => {
+                let block = builder.create_block();
+                dest_block_sequence.push((default as usize, block));
+                *entry.insert(block)
+            }
+        };
+        builder.ins().switch(val, data, default_branch_block, span);
+        for (depth, dest_block) in dest_block_sequence {
+            builder.switch_to_block(dest_block);
+            builder.seal_block(dest_block);
+            let real_dest_block = {
+                let i = state.control_stack.len() - 1 - depth;
+                let frame = &mut state.control_stack[i];
+                frame.set_branched_to_exit();
+                frame.br_destination()
+            };
+            let destination_args = state.peekn_mut(return_count);
+            builder.ins().br(real_dest_block, destination_args, span);
+        }
+        state.popn(return_count);
+    }
+    state.reachable = false;
+    Ok(())
+}
+
+/// Return the total Miden VM memory size (2^32 addresses * word (4 felts) bytes) in 64KB pages
+const fn mem_total_pages() -> i32 {
+    let bytes_fit_in_felt = 4; // although more than 32 bits can fit into a felt, use 32 bits to be safe
+    let felts_in_word = 4;
+    let total_addresses = u32::MAX as i64;
+    (total_addresses * (felts_in_word * bytes_fit_in_felt) / (64 * 1024)) as i32
+}
+
+fn translate_load(
+    ptr_ty: Type,
+    memarg: &MemArg,
+    state: &mut FuncTranslationState,
+    builder: &mut FunctionBuilderExt<'_>,
+    span: SourceSpan,
+) {
+    let addr_int = state.pop1();
+    let addr = prepare_addr(addr_int, &ptr_ty, memarg, builder, span);
+    state.push1(builder.ins().load(addr, span));
+}
+
+fn translate_load_sext(
+    ptr_ty: Type,
+    sext_ty: Type,
+    memarg: &MemArg,
+    state: &mut FuncTranslationState,
+    builder: &mut FunctionBuilderExt<'_>,
+    span: SourceSpan,
+) {
+    let addr_int = state.pop1();
+    let addr = prepare_addr(addr_int, &ptr_ty, memarg, builder, span);
+    let val = builder.ins().load(addr, span);
+    let sext_val = builder.ins().sext(val, sext_ty, span);
+    state.push1(sext_val);
+}
+
+fn translate_load_zext(
+    ptr_ty: Type,
+    zext_ty: Type,
+    memarg: &MemArg,
+    state: &mut FuncTranslationState,
+    builder: &mut FunctionBuilderExt<'_>,
+    span: SourceSpan,
+) {
+    let addr_int = state.pop1();
+    let addr = prepare_addr(addr_int, &ptr_ty, memarg, builder, span);
+    let val = builder.ins().load(addr, span);
+    let sext_val = builder.ins().zext(val, zext_ty, span);
+    state.push1(sext_val);
+}
+
+fn translate_store(
+    ptr_ty: Type,
+    memarg: &MemArg,
+    state: &mut FuncTranslationState,
+    builder: &mut FunctionBuilderExt<'_>,
+    span: SourceSpan,
+) {
+    let (addr_int, val) = state.pop2();
+    let val_ty = builder.func().dfg.value_type(val);
+    let arg = if ptr_ty != *val_ty {
+        builder.ins().trunc(val, ptr_ty.clone(), span)
```

Comment:
This will fail if `val_ty` is larger than `ptr_ty`, as truncation must be to a smaller type (it's allowed for the types to be the same size, in which case it is a no-op, but better to cast the pointer type in that case)

Reply:
I'm confused. When called for Wasm `i32_store8` op `val_ty` is `I32` and `ptr_ty` is `U8`. Why will it fail to `trunc` `I32` value to `U8`? Is it because I need to trunc to ` I8` first and then cast to `U8`?

Reply:
I think what caught my eye here was that there doesn't seem to be anything that ensures `ptr_ty` is smaller than `val_ty` - I thought I had observed one place where that could be the case, but I may have been mistaken on that. In any case, there is an assertion in the builder, so if it ever happens we'll know about it, so I think we can call this one resolved.

Reply:
Oh, I see. All store ops in Wasm have `ptr_ty` smaller or equal to `val_ty`. That's why I have not thought about adding such a check.